### ISSUE TITLE
Cannot use variable values as slot names in loop

### ISSUE THREAD

**Issue by spacedawwwg** [withastro/compiler#868]
### Astro Info

```block
Astro                    v3.0.10
Node                     v20.5.1
System                   Linux (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/vue
```

### Describe the Bug

I have tried to use variable values as slot names but it isn't working (see below).

NOTE: I've simplified my use case here for code example only - I do have a valid use case for sending HTML to slots from outside the Example component (tables with HTML content)

This is a simple Vue component using named slots based on values in the props.

```Vue
<script lang="ts" setup>
export interface Props {
  items: {
    id: string;
  }[];
}

defineProps<Props>();
</script>

<template>
    <template v-for="item of items">
      <div
        v-if="$slots && $slots[item.id]"
      >
        <slot :name="item.id" />
      </div>
    </template>
</template>
```

This is an example of me using the above vue component within an astro file to try populate slots with a mapped value.

```astro
---
const items = [
  { id: "1"},
  { id: "2"},
  { id: "abc"}
];
---
<Example items={items}>
  {items.map(({ id }) => <p slot={id}>{id}</p>)}
</Example>
```
Unfortunately I'm being hit with:

<img width="943" alt="image" src="https://github.com/withastro/astro/assets/1691867/74ab605e-81bb-4dd5-83af-ac97e1325b9a">

I have logged the `id` and it it is defined... just not when used as the slot name

### What's the expected result?

I'd expect astro to use the value to populate the named slot in the child component. with the output:

```HTML
<div>
 <p>1</p>
</div>
<div>
 <p>2</p>
</div>
<div>
 <p>abc</p>
</div>
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-vqaya3

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by spacedawwwg** [comment]
> If you paste the original example in https://live-astro-compiler.vercel.app, you can see it's generated as (formatted for readability):
> 
> ```js
> const items = [{ id: '1' }, { id: '2' }, { id: 'abc' }]
> 
> return $$render`${$$renderComponent(
>   $$result,
>   'Example',
>   Example,
>   { items: items },
>   {
>     [id]: () =>
>       $$render`${items.map(
>         ({ id }) => $$render`${$$maybeRenderHead($$result)}<p>${id}</p>`
>       )}`
>   }
> )}`
> ```
> 
> The 5th parameter is the slots object. Looks like it's not generating the slots right. I'm not sure if it's possible to fix this in the compiler, but I'll move it there at the mean time as it's a compiler issue.

yeah, so it feels like while processing the map in the compiler, if it detects slot attributes break up the map into something like:

```js
const items = [{ id: '1' }, { id: '2' }, { id: 'abc' }]

return $$render`${$$renderComponent(
  $$result,
  'Example',
  Example,
  { items: items },
  {
    ['1']: () => $$render`${$$maybeRenderHead($$result)}<p>1</p>`,
    ['2']: () => $$render`${$$maybeRenderHead($$result)}<p>2</p>`,
    ['abc']: () => $$render`${$$maybeRenderHead($$result)}<p>abc</p>`
  }
)}`
```

**Comment by MoustaphaDev** [comment]
I'm not sure we can fix this. In this case, the key of the slot render function is derived from a value defined inside the scope of the `map` function, and the compiler can't really tap into the `items` object and generate render functions of for each value like in your last comment.

**Comment by MoustaphaDev** [comment]
Reopening this as we've reverted the changes made in https://github.com/withastro/compiler/pull/963, due to unexpected regressions that it caused

**Comment by spacedawwwg** [comment]
Any movement on this @MoustaphaDev ?

**Comment by sangrepura** [comment]
LGTM

**Comment by Princesseuh** [comment]
Hello, the Go compiler for Astro is no longer maintained. All future development will be focused on [our new Rust compiler](https://github.com/withastro/compiler-rs). If this issue is still relevant and you can reproduce it on the Rust compiler, please open a new issue there. Thank you!
